В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [ ]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [ ]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [ ]:
mean_heights = np.mean(heights)
sd_height = np.std(heights, ddof = 1)
alpha = 0.05
null_height = 170
sample_size = len(heights)

In [ ]:
ztest_Score, p_value = ztest(heights, value = mean_heights, alternative='larger')

In [ ]:
ztest_Score, p_value

(np.float64(0.0), np.float64(0.5))

In [ ]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**Спостереження**

ztest_Score = 0 означає, що середнє значення вибірки точно дорівнює гіпотетичному середньому, тобто 170 см.

p_value = 0.5 свідчить про повну відсутність підстав вважати, що вибіркове середнє більше за 170.

Тобто, вибірка heights не відрізняється від очікуваного середнього.

**Висновок**

На рівні значущості 0.05 немає статистично значущих доказів, що середній ріст студентів у вибірці істотно більший за 170 см.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [ ]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [ ]:
np.mean(group_a_scores), np.mean(group_b_scores)

(np.float64(77.08125), np.float64(74.73875000000001))

In [ ]:
ztest_Score, p_value = ztest(group_a_scores, group_b_scores, alternative='two-sided')

In [ ]:
ztest_Score, p_value

(np.float64(2.357354864751502), np.float64(0.01840564855252801))

In [ ]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


**Висновок**

Оскільки p_value = 0.0184 і воно менше за рівень значущості, яке дорінює 0.05, ми відхиляємо нульову гіпотезу.

Це означає що існує статистично значуща різниця між середніми оцінками двох груп студентів. Припущення, що новий курс ефективніший, справдилось на рівні значущості 0.05.
У цьому випадку, з таким низьким p-значенням, ми маємо достатньо доказів, щоб стверджувати, що новий навчальний курс дійсно призвів до статистично кращих результатів для Групи A порівняно зі стандартним курсом для Групи B.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [ ]:
from scipy import stats

In [ ]:
group_a_scores_small = group_a_scores[:5]

In [ ]:
group_a_scores_small

array([78.55, 72.25, 79.88, 75.  , 76.54])

In [ ]:
ttest_Score, p_value = stats.ttest_ind(group_a_scores_small, group_b_scores)

In [ ]:
ttest_Score, p_value

(np.float64(0.8167951030410814), np.float64(0.4185479477535733))

In [ ]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


**Висновок**

При використанні лише перших 5 записів з Групи А, p_value t-тесту становить 0.4185.

Оскільки p_value більше ніж рівень значущості, ми не можемо відхилити нульову гіпотезу. Це означає, що, незважаючи на потенційну різницю в середніх значеннях, з такою малою вибіркою для Групи А немає достатніх статистичних доказів, щоб стверджувати про значущу різницю між середніми оцінками цих двох груп на рівні значущості 0.05. Вибірка з 5 учнів надто мала, щоб зробити надійний висновок про ефективність нового курсу порівняно з усією групою В. Цей результат доводить, як розмір вибірки впливає на можливість виявлення статистично значущих відмінностей. Менші вибірки мають меншу статистичну потужність.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [ ]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [ ]:
# Перетворюємо списки на numpy масиви для зручності
before_training = np.array(before_training)
after_training = np.array(after_training)

In [ ]:
# Обчислюємо різниці для кожного працівника (після - до)
differences = after_training - before_training

In [ ]:
differences

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

In [ ]:
np.mean(differences)

np.float64(3.1499999999999995)

In [ ]:
ztest_Score, p_value = ztest(differences, value=0, alternative='larger')

In [ ]:
ztest_Score, p_value

(np.float64(9.13890009623916), np.float64(3.1545359919175505e-20))

In [ ]:
print(f"p_value: {p_value:.4f}")

p_value: 0.0000


In [ ]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Відхилити Н0.


**Висновок**

Оскільки отримане p-p_value є надзвичайно малим і значно меншим за рівень значущості 0.05, ми маємо відхилити нульову гіпотезу.

Отже, існує статистично значуще покращення показників продуктивності працівників після проходження тренінгу. Висока ztest_Score (9.1389) додатково підтверджує цей сильний ефект. Можна впевнено стверджувати, що тренінг мав позитивний вплив на продуктивність.